rust/agents/updater/src/updater.rs

```diff
@@ -107,7 +107,14 @@ impl NomadAgent for Updater {
             self.submitted_update_count.clone(),
         );
 
+        let fail_check = self.is_home_failed();
+        let home_fail_watch_task = self.watch_home_fail(self.interval_seconds);
+
         tokio::spawn(async move {
+            if fail_check.await?? {
```

Comment:
> `if fail_check.await??`

This is **not** a safe assumption. The fail watch task could error out for another reason (for example when making RPC call `home.state()`). I would make sure that the error is of type `BaseError::FailedHome` before bubbling up a failed home error

Reply:
Don't want to start crash looping with failed home errors and scare everyone if its really just an RPC issue or something else

Reply:
The `fn is_home_failed(&self) -> Instrumented<JoinHandle<Result<bool>>>` method returns `true`, only if it correctly got the state **and** the state equals to `nomad_core::State::Failed`, otherwise it returns the RPC error. Hence, the check that u highlighted:
1. With the first `?` bubbles tokio's await issue if it exists, then
2. The second `?` bubbles the possible RPC issue from `.is_home_failed()`, 
3. We end up with the boolean, which just tells if the home is in the failed state
4. Finally, if it is the case we return the report of the home is in the failed state

Reply:
Ok, I just saw the comment below and there is an understandable confusion, that must be addressed. There is another method that I implemented `fn watch_home_fail(&self, interval: u64) -> Instrumented<JoinHandle<Result<()>>>`. It continuously runs, until either the RPC issue happens or Home is failed, then it returns Report (`BaseError::FailedHome`). This approach works fine in `Updater` or `Processor`, because there are several other loops that just shutdown an agent on the first resolved `Future`, no matter error or not. Though in Processor, we want to watch the same method and do kind of the same - watch home, but now we need to distinguish between RPC error and failed home

Reply:
The only solution I see is to
1. make the method `watch_home_fail` which will only resolve on failed state with Ok(()), and not with the report. 
2. Watcher will react on Ok and unenroll replicas. Anything else like RPC stuff will bubble up.
3. `Updater` and `Replica` will either be calling the same method, but wrapped `()` into like `and_then(|_| Err(BaseError::FailedHome.into()))`

Q: Why don't I want to return state when `Home` is failed? A: Because we check same variable twice, which is uglier than wrapping `Ok(())` into `Err(Report)`

Reply:
Resolved at [df5f7c0](https://github.com/nomad-xyz/nomad-monorepo/pull/145/commits/df5f7c07e47004471dd3bf3e9b2ce76107baa627)